commons/all/src/main/java/org/infinispan/commons/io/FileWatcher.java

```diff
@@ -0,0 +1,116 @@
+package org.infinispan.commons.io;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+
+/**
+ * @since 15.0
+ */
+public class FileWatcher implements Runnable {
+   private static final Log log = LogFactory.getLog(FileWatcher.class);
+   public static final int SLEEP = 2000;
+   private static FileWatcher INSTANCE;
+   private final Thread thread;
+   private final ConcurrentHashMap<Path, Watched> watched;
+   private volatile boolean running = true;
+
+   public static synchronized void watch(Path path, Consumer<Path> callback) throws IOException {
+      if (INSTANCE == null) {
+         // Initialize the singleton
+         INSTANCE = new FileWatcher();
+      }
+      INSTANCE.watch0(path, callback);
+   }
+
+   public static synchronized void unwatch(Path path) {
+      if (INSTANCE == null) {
+         return;
+      }
+      INSTANCE.unwatch0(path);
+   }
+
+   private void unwatch0(Path path) {
+      watched.remove(path);
+      log.debugf("Unwatched %s", path);
+   }
+
+   private void watch0(Path path, Consumer<Path> callback) {
+      watched.compute(path, (k, w) -> {
+         if (w == null) {
+            w = new Watched();
+            try {
+               w.lastModified = Files.getLastModifiedTime(path).toMillis();
+            } catch (FileNotFoundException e) {
+               w.lastModified = -1;
+               log.debugf("File not found %s", path);
+            } catch (IOException e) {
+               throw new RuntimeException(e);
+            }
+         }
+         w.watchers.add(callback);
+         return w;
+      });
+      log.debugf("Watching %s", path);
+   }
+
+   private FileWatcher() {
+      watched = new ConcurrentHashMap<>();
+      thread = new Thread(this, "FileWatcher");
+      Runtime.getRuntime().addShutdownHook(new Thread(this::stop));
+      thread.start();
+   }
+
+   @Override
+   public void run() {
+      while (running) {
+         try {
+            Thread.sleep(SLEEP);
```

Comment:
Shouldn't we be making use of a `ScheduledExecutorService` and should be creating this via `NamedExecutorsFactory`?